### ISSUE TITLE
[Williams %R超买超卖策略] | Williams %R Overbought Oversold

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4496]
**标签**: quantitative-strategy
### 摘要
基于Williams %R指标的经典超买超卖判别，结合自适应阈值和多重时间框架共振，构建趋势反转交易系统。

### 算法描述

**核心原理**

Williams %R（威廉指标）衡量收盘价相对于特定周期内价格区间的位置，以-100至0表示。当指标进入超买区（>-20）或超卖区（<-80）时，市场可能面临回调压力。

**数学模型**

1. **Williams %R计算**：
   $$W\%R = \frac{H_n - C_n}{H_n - L_n} \times (-100)$$
   其中 $H_n$ 为n周期最高价，$L_n$ 为n周期最低价，$C_n$ 为当前收盘价。

2. **EMA平滑版本**：
   $$W\%R_{smooth} = EMA(W\%R, k)$$
   使用k周期指数移动平均减少假信号。

3. **自适应阈值**：
   $$Threshold_{OB} = -20 + \alpha \cdot \sigma_{WR}$$
   $$Threshold_{OS} = -80 - \beta \cdot \sigma_{WR}$$
   其中 $\sigma_{WR}$ 为指标标准差，$\alpha, \beta$ 为可调参数。

4. **多重时间框架共振**：
   $$Signal = \prod_{t \in \{15m, 1h, 4h, 1d\}} \text{Signal}_t$$
   多周期同向信号时胜率显著提升。

5. **动量确认**：
   $$Momentum = \frac{W\%R_t - W\%R_{t-n}}{n}$$
   动量方向辅助过滤假突破。

6. **布林带过滤**：
   $$Filter = \frac{C - MA_{20}}{2 \cdot \sigma_{20}}$$
   价格位于布林带中轨附近时信号更可靠。

**策略流程**：
1. 计算多周期Williams %R（15m/1h/4h/1D）
2. 识别超买超卖区域的交叉信号
3. 等待价格触及布林带边缘确认
4. 检查动量方向过滤噪音
5. 计算风险回报比入场
6. 追踪止损保护利润

### Python代码示例

```python
import numpy as np
import pandas as pd
from typing import Dict, List, Optional, Tuple
from collections import deque

class WilliamsPercentR:
    """Williams %R指标计算器"""
    
    def __init__(self, period: int = 14):
        self.period = period
        
    def calculate(self, highs: pd.Series, lows: pd.Series, 
                 closes: pd.Series) -> pd.Series:
        """
        计算Williams %R
        W%R = (H_n - C) / (H_n - L_n) * (-100)
        """
        highest_high = highs.rolling(window=self.period).max()
        lowest_low = lows.rolling(window=self.period).min()
        
        wr = -100 * (highest_high - closes) / (highest_high - lowest_low + 1e-10)
        return wr
    
    def calculate_ema_smoothed(self, wr: pd.Series, ema_period: int = 5) -> pd.Series:
        """计算EMA平滑版本"""
        return wr.ewm(span=ema_period, adjust=False).mean()
    
    def calculate_momentum(self, wr: pd.Series, n: int = 5) -> pd.Series:
        """计算动量"""
        return wr.diff(n)


class WilliamsRStrategy:
    """Williams %R超买超卖策略"""
    
    def __init__(self):
        self.timeframes = {
            '15m': {'period': 14, 'ema_period': 5, 'weight': 1.0},
            '1h': {'period': 14, 'ema_period': 5, 'weight': 2.0},
            '4h': {'period': 14, 'ema_period': 5, 'weight': 3.0},
            '1d': {'period': 14, 'ema_period': 5, 'weight': 4.0}
        }
        
        # 超买超卖阈值
        self.ob_threshold = -20
        self.os_threshold = -80
        
        # 自适应参数
        self.alpha_ob = 0.5
        self.alpha_os = 0.5
        
        # 布林带参数
        self.bb_period = 20
        self.bb_std = 2
        
        # 交易参数
        self.risk_reward_ratio = 2.0
        self.max_loss_pct = 0.02
        
        # 历史数据
        self.price_history = deque(maxlen=200)
        self.signals_history = []
        
    def calculate_adaptive_thresholds(self, wr_series: pd.Series) -> Tuple[float, float]:
        """计算自适应阈值"""
        recent_wr = wr_series.dropna().tail(60)
        
        if len(recent_wr) < 20:
            return self.ob_threshold, self.os_threshold
            
        std_wr = recent_wr.std()
        
        # 动态调整阈值
        ob = self.ob_threshold + self.alpha_ob * std_wr
        os = self.os_threshold - self.alpha_os * std_wr
        
        return ob, os
    
    def detect_crossover(self, wr: pd.Series, threshold: float) -> List[Dict]:
        """检测交叉信号"""
        signals = []
        
        wr_values = wr.dropna().values
        dates = wr.dropna().index
        
        for i in range(1, len(wr_values)):
            # 上穿（从下往上突破超卖区）
            if wr_values[i-1] < threshold and wr_values[i] >= threshold:
                signals.append({
                    'date': dates[i],
                    'type': 'oversold_cross_up',
                    'value': wr_values[i],
                    'threshold': threshold
                })
            # 下穿（从上往下突破超买区）
            elif wr_values[i-1] > threshold and wr_values[i] <= threshold:
                signals.append({
                    'date': dates[i],
                    'type': 'overbought_cross_down',
                    'value': wr_values[i],
                    'threshold': threshold
                })
                
        return signals
    
    def calculate_bollinger_filter(self, prices: pd.Series) -> pd.Series:
        """计算布林带过滤信号"""
        ma = prices.rolling(window=self.bb_period).mean()
        std = prices.rolling(window=self.bb_period).std()
        
        upper_band = ma + self.bb_std * std
        lower_band = ma - self.bb_std * std
        
        # 价格位置（0-1之间）
        bb_position = (prices - lower_band) / (upper_band - lower_band + 1e-10)
        
        return bb_position
    
    def check_timeframe_resonance(self, wr_signals: Dict[str, List[Dict]],
                                  current_date, tolerance_hours: int = 4) -> Dict:
        """
        检查多重时间框架共振
        多周期同向信号时胜率显著提升
        """
        resonance_score = 0
        timeframe_weights = {'15m': 1, '1h': 2, '4h': 3, '1d': 4}
        
        # 检查每个时间框架的信号
        for tf, signals in wr_signals.items():
            # 找到最近的信号
            recent = [s for s in signals 
                     if abs((current_date - s['date']).total_seconds() / 3600) < tolerance_hours * timeframe_weights.get(tf, 1)]
            
            if recent:
                if 'oversold_cross_up' in recent[0]['type']:
                    resonance_score += timeframe_weights.get(tf, 1)
                elif 'overbought_cross_down' in recent[0]['type']:
                    resonance_score -= timeframe_weights.get(tf, 1)
        
        # 归一化
        max_score = sum(self.timeframes[tf]['weight'] for tf in self.timeframes)
        resonance_ratio = resonance_score / max_score if max_score > 0 else 0
        
        return {
            'resonance_score': resonance_score,
            'resonance_ratio': resonance_ratio,
            'signal_direction': 'LONG' if resonance_score > 0 else 'SHORT' if resonance_score < 0 else 'NEUTRAL'
        }
    
    def calculate_entry_price(self, price: float, signal_type: str,
                             bb_position: float) -> Tuple[float, float]:
        """
        计算入场价格和止损
        """
        if signal_type == 'LONG':
            # 买入信号：价格接近布林下轨
            entry = price * 0.998  # 略低于现价
            stop_loss = price * (1 - self.max_loss_pct)
        else:
            # 卖出信号：价格接近布林上轨
            entry = price * 1.002
            stop_loss = price * (1 + self.max_loss_pct)
        
        return entry, stop_loss
    
    def calculate_position_size(self, entry_price: float, 
                               stop_loss: float, 
                               total_capital: float) -> int:
        """计算仓位大小"""
        risk_amount = total_capital * self.max_loss_pct
        price_risk = abs(entry_price - stop_loss)
        
        if price_risk == 0:
            return 0
            
        position_size = int(risk_amount / price_risk)
        return min(position_size, int(total_capital * 0.1 / entry_price))
    
    def generate_signal(self, multi_tf_data: Dict[str, pd.DataFrame],
                       current_date: datetime) -> Optional[Dict]:
        """生成综合交易信号"""
        # 计算每个时间框架的Williams %R
        wr_calculator = WilliamsPercentR(period=14)
        
        wr_signals = {}
        current_wr = {}
        
        for tf, df in multi_tf_data.items():
            wr = wr_calculator.calculate(df['high'], df['low'], df['close'])
            wr_smooth = wr_calculator.calculate_ema_smoothed(wr)
            
            # 计算自适应阈值
            ob, os = self.calculate_adaptive_thresholds(wr_smooth)
            
            # 检测交叉
            signals = []
            signals.extend(self.detect_crossover(wr_smooth, ob))
            signals.extend(self.detect_crossover(wr_smooth, os))
            
            wr_signals[tf] = signals
            current_wr[tf] = wr_smooth.iloc[-1] if len(wr_smooth) > 0 else -50
        
        # 检查共振
        resonance = self.check_timeframe_resonance(wr_signals, current_date)
        
        # 布林带过滤
        bb_pos = self.calculate_bollinger_filter(multi_tf_data['1h']['close'])
        current_bb_pos = bb_pos.iloc[-1] if len(bb_pos) > 0 else 0.5
        
        # 动量确认
        momentum = wr_calculator.calculate_momentum(wr_smooth, n=5)
        current_momentum = momentum.iloc[-1] if len(momentum) > 0 else 0
        
        # 信号生成
        signal = None
        
        if resonance['resonance_ratio'] > 0.5 and current_bb_pos < 0.2:
            # 多周期共振做多，且价格接近布林下轨
            entry, stop = self.calculate_entry_price(
                multi_tf_data['1h']['close'].iloc[-1], 'LONG', current_bb_pos
            )
            
            signal = {
                'direction': 'LONG',
                'entry_price': entry,
                'stop_loss': stop,
                'take_profit': entry + (entry - stop) * self.risk_reward_ratio,
                'confidence': resonance['resonance_ratio'],
                'timeframe_resonance': resonance,
                'reason': 'multi_tf_oversold共振'
            }
            
        elif resonance['resonance_ratio'] < -0.5 and current_bb_pos > 0.8:
            entry, stop = self.calculate_entry_price(
                multi_tf_data['1h']['close'].iloc[-1], 'SHORT', current_bb_pos
            )
            
            signal = {
                'direction': 'SHORT',
                'entry_price': entry,
                'stop_loss': stop,
                'take_profit': entry - (stop - entry) * self.risk_reward_ratio,
                'confidence': abs(resonance['resonance_ratio']),
                'timeframe_resonance': resonance,
                'reason': 'multi_tf_overbought共振'
            }
        
        return signal
    
    def backtest_strategy(self, data: pd.DataFrame, 
                         initial_capital: float = 100000) -> Dict:
        """回测策略"""
        capital = initial_capital
        position = None
        trades = []
        
        # 计算Williams %R
        wr_calculator = WilliamsPercentR(period=14)
        wr = wr_calculator.calculate(data['high'], data['low'], data['close'])
        wr_smooth = wr_calculator.calculate_ema_smoothed(wr)
        
        # 计算布林带
        bb_pos = self.calculate_bollinger_filter(data['close'])
        
        # 自适应阈值
        ob, os = self.calculate_adaptive_thresholds(wr_smooth)
        
        for i in range(50, len(data)):
            current_date = data.index[i]
            current_price = data['close'].iloc[i]
            current_wr = wr_smooth.iloc[i]
            current_bb = bb_pos.iloc[i]
            
            # 入场逻辑
            if position is None:
                # 做多条件
                if current_wr < os and current_bb < 0.2:
                    entry_price = current_price * 0.998
                    stop_loss = current_price * (1 - self.max_loss_pct)
                    size = self.calculate_position_size(entry_price, stop_loss, capital)
                    
                    if size > 0:
                        position = {
                            'entry_date': current_date,
                            'entry_price': entry_price,
                            'stop_loss': stop_loss,
                            'size': size,
                            'direction': 'LONG'
                        }
                        
                # 做空条件
                elif current_wr > ob and current_bb > 0.8:
                    entry_price = current_price * 1.002
                    stop_loss = current_price * (1 + self.max_loss_pct)
                    size = self.calculate_position_size(entry_price, stop_loss, capital)
                    
                    if size > 0:
                        position = {
                            'entry_date': current_date,
                            'entry_price': entry_price,
                            'stop_loss': stop_loss,
                            'size': size,
                            'direction': 'SHORT'
                        }
            else:
                # 出场逻辑
                should_exit = False
                
                if position['direction'] == 'LONG':
                    if current_price <= position['stop_loss']:
                        should_exit = True
                        pnl = (current_price - position['entry_price']) * position['size']
                    elif current_wr > ob:  # 超买区平仓
                        should_exit = True
                        pnl = (current_price - position['entry_price']) * position['size']
                else:
                    if current_price >= position['stop_loss']:
                        should_exit = True
                        pnl = (position['entry_price'] - current_price) * position['size']
                    elif current_wr < os:  # 超卖区平仓
                        should_exit = True
                        pnl = (position['entry_price'] - current_price) * position['size']
                
                if should_exit:
                    capital += pnl
                    trades.append({**position, 'exit_date': current_date, 
                                  'exit_price': current_price, 'pnl': pnl})
                    position = None
        
        return {
            'final_capital': capital,
            'total_return': (capital - initial_capital) / initial_capital,
            'num_trades': len(trades),
            'win_rate': sum(1 for t in trades if t['pnl'] > 0) / max(len(trades), 1),
            'avg_win': np.mean([t['pnl'] for t in trades if t['pnl'] > 0]) if trades else 0,
            'avg_loss': np.mean([t['pnl'] for t in trades if t['pnl'] <= 0]) if trades else 0
        }


if __name__ == "__main__":
    # 测试
    np.random.seed(42)
    dates = pd.date_range('2024-01-01', periods=200, freq='1h')
    data = pd.DataFrame({
        'high': 100 + np.random.randn(200).cumsum() + 5,
        'low': 100 + np.random.randn(200).cumsum() - 5,
        'close': 100 + np.random.randn(200).cumsum()
    }, index=dates)
    
    strategy = WilliamsRStrategy()
    
    wr_calculator = WilliamsPercentR(period=14)
    wr = wr_calculator.calculate(data['high'], data['low'], data['close'])
    wr_smooth = wr_calculator.calculate_ema_smoothed(wr)
    
    print(f"Williams %R 范围: [{wr_smooth.min():.2f}, {wr_smooth.max():.2f}]")
    print(f"超买阈值: {strategy.ob_threshold}, 超卖阈值: {strategy.os_threshold}")
